java/core/src/test/org/apache/orc/TestOrcTimestampPPD.java

```diff
@@ -0,0 +1,173 @@
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TestName;
+
+import java.io.File;
+import java.io.IOException;
+import java.sql.Timestamp;
+import java.util.List;
+import java.util.TimeZone;
+
+import static junit.framework.Assert.assertEquals;
+
+public class TestOrcTimestampPPD {
+  Path workDir =
+      new Path(System.getProperty("test.tmp.dir", "target" + File.separator + "test" + File.separator + "tmp"));
+  Configuration conf;
+  FileSystem fs;
+  Path testFilePath;
+  static TimeZone defaultTimeZone = TimeZone.getDefault();
+
+  public TestOrcTimestampPPD() {
+  }
+
+  @Rule
+  public TestName testCaseName = new TestName();
+
+  @Before
+  public void openFileSystem() throws Exception {
+    conf = new Configuration();
+    fs = FileSystem.getLocal(conf);
+    testFilePath = new Path(workDir, "TestOrcTimestampPPD." + testCaseName.getMethodName() + ".orc");
+    fs.delete(testFilePath, false);
+  }
+
+  @After
+  public void restoreTimeZone() {
+    TimeZone.setDefault(defaultTimeZone);
+  }
+
+  @Test
+  // ORC-611 : PPD evaluation with min-max stats for sub-millisecond timestamps
+  public void testSubMsTimestampWriterStats() throws Exception {
+    TypeDescription schema = TypeDescription.createTimestamp();
+    TimeZone.setDefault(TimeZone.getTimeZone("America/Los_Angeles"));
+
+    Writer writer = OrcFile.createWriter(testFilePath,
+        OrcFile.writerOptions(conf).setSchema(schema).stripeSize(100000).bufferSize(10000)
+            .version(OrcFile.Version.CURRENT));
+
+    List<Timestamp> tslist = Lists.newArrayList();
+    tslist.add(Timestamp.valueOf("1970-01-01 00:00:00.0005"));
+
+    VectorizedRowBatch batch = schema.createRowBatch();
+    TimestampColumnVector times = (TimestampColumnVector) batch.cols[0];
+    for (Timestamp t : tslist) {
+      times.set(batch.size++, t);
+    }
+    times.isRepeating = true;
+    writer.addRowBatch(batch);
+    // Done writing to file
+    writer.close();
+
+    TimeZone.setDefault(TimeZone.getTimeZone("America/Los_Angeles"));
+    // Now reading
+    Reader reader = OrcFile.createReader(testFilePath, OrcFile.readerOptions(conf).filesystem(fs));
+
+    RecordReader rows = reader.rows();
+    batch = reader.getSchema().createRowBatch();
+    times = (TimestampColumnVector) batch.cols[0];
+    while (rows.nextBatch(batch)) {
+      for (int r = 0; r < batch.size; ++r) {
+        System.out.println("ts: "+ times.asScratchTimestamp(r));
+        Assert.assertEquals(tslist.get(0), times.asScratchTimestamp(r));
+        Assert.assertEquals(tslist.get(0).getNanos(), times.asScratchTimestamp(r).getNanos());
+      }
+    }
+    rows.close();
+    ColumnStatistics[] colStats = reader.getStatistics();
+    Timestamp gotMin = ((TimestampColumnStatistics) colStats[0]).getMinimum();
+    assertEquals("1970-01-01 00:00:00.0", gotMin.toString());
+
+    Timestamp gotMax = ((TimestampColumnStatistics) colStats[0]).getMaximum();
```

Comment:
A comment could explain that this is a bit buggy in the writer, but the reader can handle it.

Reply:
A proper solution, in general, would include lower and upper bounds similar to what we do for string shortening (https://issues.apache.org/jira/browse/ORC-203) -- see Owens reply above.

To achieve that we cannot really avoid bumping the writer version so I will update the patch accordingly

Reply:
@csringhofer please let me know what you think about the latest changes!